### PR TITLE
Miscellaneous packaging improvements

### PR DESCRIPTION
**Pull request by khaeru**
- Use pyproject.toml for packaging, instead of setup.{cfg,py}.
- Replace usage of deprecated `pkg_resources` with `importlib.metadata`/`importlib_metadata` (Python 3.7 only).
- Various housekeeping and clean-up of the CI workflows.
- Use `pytest-xdist` to parallelize tests (n=2) on GHA runners; this shortens run time of individual jobs by up to 5 minutes.
- Cherry-pick two commits (by @glatterf42 originally at #494) adjusting for pandas 2.0.0.

## How to review

Read the diff and note that the CI checks all pass.

## PR checklist

- [x] Merge iiasa/ixmp#476, required to address an issue with the codecov dependency.
- [x] Continuous integration checks all ✅
- [x] ~Add or expand tests;~ coverage checks both ✅
- ~Add, expand, or update documentation.~ N/A
- ~Update release notes.~

### CODE REVIEW

**Comment by khaeru** [comment]
@glatterf42 #683 led to conflicts in test_tutorials.py; please wait for me to rebase &  deconflict before reviewing.

**Comment by glatterf42** [comment]
Sorry, I didn't realise it would. I'll review once you are done.

File: .github/workflows/pytest.yaml [hunk 1]
```diff
@@ -108,13 +90,13 @@ jobs:
...
-      run: pip install .
-
     - name: Install Python package and dependencies
-      working-directory: message_ix
-      run: pip install .[tests]
+      # By default, the below installs ixmp from the main branch. To run against
+      # other code, e.g. other branches for open PRs), temporarily edit as
+      # appropriate. DO NOT merge such changes to `main`.
+      run: |
+        pip install --upgrade "ixmp @ git+https://github.com/iiasa/ixmp.git@main"
```
Comment (glatterf42) [inline .github/workflows/pytest.yaml:98]: Out of curiosity: 'edit as appropriate' means to replace 'main' in the command with the name of the branch that should be tested?

File: message_ix/tests/test_tutorials.py [hunk 2]
```diff
@@ -1,85 +1,109 @@
...
 ]
 
 # Short, readable IDs for the tests. Use getattr() to unpack the values from
 # pytest.param()
-ids = [getattr(arg, "values", arg)[0][-1] for arg in tutorials]
+IDS = [getattr(arg, "values", arg)[0][-1] for arg in TUTORIALS]
 
-# List of data files required to run tutorials
-data_files = [
+#: List of data files required by "westeros_baseline_using_xlsx_import_part2".
```
Comment (glatterf42) [inline message_ix/tests/test_tutorials.py:105]: What is the purpose of this `:`?

File: message_ix/tests/test_tutorials.py [hunk 3]
```diff
@@ -97,27 +121,26 @@ def nb_path(request, tutorial_path):
 
 # Parametrize the first 3 arguments using the variables *tutorial* and *ids*.
```
Comment (glatterf42) [inline message_ix/tests/test_tutorials.py:122]:
```suggestion
# Parametrize the first 3 arguments using the variables *TUTORIALS* and *IDS*.
```
These are capitalized now.

File: pyproject.toml [hunk 4]
```diff
@@ -1,116 +1,105 @@
...
+  'importlib_metadata; python_version < "3.8"',
+  "ixmp >= 3.5.0",
+  "numpy",
+  "pandas >= 1.2",
+  "PyYAML",
+]
+
+[project.urls]
+homepage = "http://github.com/iiasa/message_ix"
+repository = "http://github.com/iiasa/message_ix"
```
Comment (glatterf42) [inline pyproject.toml:44]:
```suggestion
repository = "https://github.com/iiasa/message_ix"
```

File: pyproject.toml [hunk 5]
```diff
@@ -1,116 +1,105 @@
...
+  "click",
+  'importlib_metadata; python_version < "3.8"',
+  "ixmp >= 3.5.0",
+  "numpy",
+  "pandas >= 1.2",
+  "PyYAML",
+]
+
+[project.urls]
+homepage = "http://github.com/iiasa/message_ix"
```
Comment (glatterf42) [inline pyproject.toml:43]:
```suggestion
homepage = "https://github.com/iiasa/message_ix"
```

**Review by glatterf42 — APPROVED** [review]
Looks good to me, only one question and a few very minor remarks from my side.

File: .github/workflows/pytest.yaml (hunk 1 above)
Comment (khaeru) [reply .github/workflows/pytest.yaml:98]: Yes, exactly.

File: message_ix/tests/test_tutorials.py (hunk 2 above)
Comment (khaeru) [reply message_ix/tests/test_tutorials.py:105]: These are picked up by sphinx.ext.autodoc. See [here](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-automethod) in the para beginning "For module data members and class attributes, …"

**Comment by khaeru** [comment]
NB the codecov/patch check failure occurs because a few lines are changed in .tools.add_year, but that code remains untested. The tests will be added by #494, from which those changes were cherry-picked.

File: message_ix/tests/test_tutorials.py (hunk 2 above)
Comment (glatterf42) [reply message_ix/tests/test_tutorials.py:105]: Thanks :)

**Comment by glatterf42** [comment]
Do we now have to remove those commits from #494 or does git recognize that we included them here already once we merge both?

**Comment by khaeru** [comment]
> Do we now have to remove those commits from #494 or does git recognize that we included them here already once we merge both?

Yes, they can be removed in a few steps:

1. Check out that branch.
2. `git rebase -i HEAD~10` —replacing "10" with some number that goes back far enough to include those commits.
3. Edit the TODO list in the editor that is opened. Put "d" or "drop" on the start of the line for those commits that have been cherry-picked; leave the rest the same.
4. `git rebase main`; this will move the branch to the head of `main`, which already contains the cherry-picked commits.

Another way is to just do (1) and (4). During the rebase, git will complain, twice, that the cherry-picked commit "is (now) empty", because all changes they contained are already on `main`. It will prompt you to do `git rebase --skip` each time.

**Comment by glatterf42** [comment]
This is mostly just FYI: I've applied your suggestion to #494 and ran (1) and (4) after updating my local `main`. The process was slightly different for me, then: git immediately told me that it skipped the two previously applied commits. It still found two merge conflicts, however, which I had to resolve manually (both about the beginning of the definition of the `interpolate_1d` function). After that, I had to force push the branch because my initial updates to `main` were not yet included in the upstream branch.

**Comment by khaeru** [comment]
In this PR I forgot to adjust this line in the "nightly" CI workflow:
https://github.com/iiasa/message_ix/blob/a824be9f5a641ec4c998485c886a77ff0f39ab91/.github/workflows/nightly.yaml#L47

…to refer to pyproject.toml, mirroring the change made to the "pytest" workflow. As a result the workflow fails, e.g. https://github.com/iiasa/message_ix/actions/runs/4728750039

**Pull request merged**